tests/PolyType.Tests/IsRequiredTests.cs

```diff
@@ -0,0 +1,188 @@
+    {
+        var shape = (IObjectTypeShape?)providerUnderTest.Provider.GetShape(typeof(HasNonDefaultConstructorAndUnrelatedRequiredOverriddenProperty));
+        Assert.NotNull(shape);
+        Assert.False(shape.Properties.Single(p => p.Name == nameof(HasNonDefaultConstructorAndUnrelatedRequiredOverriddenProperty.RequiredProperty)).IsRequired);
+        Assert.True(shape.Constructor?.Parameters[0].IsRequired);
+    }
+
+    [GenerateShape]
+    public partial class HasRequiredProperty
+    {
+        public required bool RequiredProperty { get; init; }
+
+        public bool NotRequiredProperty { get; init; }
+    }
+
+    [GenerateShape]
+    public partial class HasRequiredPropertyWithNonOverriddingAttributes
+    {
+        [PropertyShape]
+        public required bool RequiredProperty { get; init; }
+
+        [PropertyShape]
+        public bool NotRequiredProperty { get; init; }
+    }
+
+    [GenerateShape]
+    public partial class HasRequiredPropertyWithOverriddingAttributes
+    {
+        [PropertyShape(IsRequired = false)]
+        public required bool RequiredProperty { get; init; }
+
+        [PropertyShape(IsRequired = true)]
+        public bool NotRequiredProperty { get; init; }
+    }
+
+    [GenerateShape]
+    public partial class HasNonDefaultConstructor
+    {
+        public HasNonDefaultConstructor(int myProperty)
+        {
+            MyProperty = myProperty;
+        }
+
+        public int MyProperty { get; set; }
+    }
+
+    [GenerateShape]
+    public partial class HasNonDefaultConstructorAndParameterNonOverride
+    {
+        public HasNonDefaultConstructorAndParameterNonOverride([ParameterShape] int myProperty)
+        {
+            MyProperty = myProperty;
+        }
+
+        public int MyProperty { get; set; }
+    }
+
+    [GenerateShape]
+    public partial class HasNonDefaultConstructorAndParameterOverride
+    {
+        public HasNonDefaultConstructorAndParameterOverride([ParameterShape(IsRequired = false)] int myProperty)
+        {
+            MyProperty = myProperty;
+        }
+
+        public int MyProperty { get; set; }
+    }
+
+    [GenerateShape]
+    public partial class HasNonDefaultConstructorWithDefaultAndRequiredProperty
+    {
+        public HasNonDefaultConstructorWithDefaultAndRequiredProperty(int myProperty = 5)
+        {
+            MyProperty = myProperty;
+        }
+
+        public required int MyProperty { get; set; }
+    }
+
+    [GenerateShape]
+    public partial class HasNonDefaultConstructorAndUnrelatedRequiredProperty
+    {
+        public HasNonDefaultConstructorAndUnrelatedRequiredProperty(int unrelated)
+        {
+        }
+
+        public required bool RequiredProperty { get; set; }
+    }
+
+    [GenerateShape]
+    public partial class HasNonDefaultConstructorAndUnrelatedRequiredOverriddenProperty
+    {
+        public HasNonDefaultConstructorAndUnrelatedRequiredOverriddenProperty(int unrelated)
+        {
+        }
```

Comment:
It's something I should have done for a while now, but would it be possible to exclude the test projects themselves from code coverage?

Reply:
I should be able to figure that out.

Reply:
#144 is my attempt.